autoprecompiles/src/range_constraint_optimizer.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::hash::Hash;
+
+use num_traits::One;
+use powdr_constraint_solver::constraint_system::{
+    BusInteraction, BusInteractionHandler, ConstraintSystem,
+};
+use powdr_constraint_solver::grouped_expression::GroupedExpression;
+use powdr_constraint_solver::inliner::DegreeBound;
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::solver::{new_solver, Solver};
+use powdr_number::FieldElement;
+
+pub type RangeConstraintMap<T, V> = BTreeMap<GroupedExpression<T, V>, RangeConstraint<T>>;
+
+pub trait RangeConstraintHandler<T: FieldElement> {
+    /// If the bus interaction *only* enforces range constraints, returns them
+    /// as a map of expressions to range constraints.
+    /// For example, if a bus interaction takes two arguments `a` and `b` and
+    /// enforces the range constraints `0 <= a < 2^b`, it is *not* a pure range
+    /// constraint if both values are unknown (because the allowed values of `a`
+    /// depend on `b`), but it is a pure range constraint if `b` is known.
```

Comment:
This sentence is not trivial to parse because it's quite long and has nested ifs without explicit precedence. I think I understood it, but I'd suggest using bullet points with indentation.

Reply:
done